zookeeper-common/src/test/java/org/apache/zookeeper/common/TestByteBufAllocator.java

```diff
@@ -0,0 +1,145 @@
+     */
+    public static TestByteBufAllocator getInstance() {
+        TestByteBufAllocator result = INSTANCE.get();
+        if (result == null) {
+            // Note: the leak detector level never gets reset after this,
+            // but that's probably ok since this is only used by test code.
+            ResourceLeakDetector.setLevel(ResourceLeakDetector.Level.PARANOID);
+            INSTANCE.compareAndSet(null, new TestByteBufAllocator());
+            result = INSTANCE.get();
+        }
+        return result;
+    }
+
+    /**
+     * Destroys the singleton testing allocator and throws an error if any of the
+     * buffers allocated by it have been leaked. Attempts to print leak details to
+     * standard error before throwing, by using netty's built-in leak tracking.
+     * Note that this might not always work, since it only triggers when a buffer
+     * is garbage-collected and calling System.gc() does not guarantee that a buffer
+     * will actually be GC'ed.
+     *
+     * This should be called at the end of a unit test's tearDown() method.
+     */
+    public static void checkForLeaks() {
+        TestByteBufAllocator result = INSTANCE.getAndSet(null);
+        if (result != null) {
+            result.checkInstanceForLeaks();
+        }
+
+    }
+
+    private final List<ByteBuf> trackedBuffers = new ArrayList<>();
+
+    public TestByteBufAllocator()
+    {
+        super(false);
+    }
+
+    @Override
+    protected ByteBuf newHeapBuffer(int initialCapacity, int maxCapacity)
+    {
+        return track(super.newHeapBuffer(initialCapacity, maxCapacity));
+    }
+
+    @Override
+    protected ByteBuf newDirectBuffer(int initialCapacity, int maxCapacity)
+    {
+        return track(super.newDirectBuffer(initialCapacity, maxCapacity));
+    }
+
+    @Override
+    public CompositeByteBuf compositeHeapBuffer(int maxNumComponents)
+    {
+        return track(super.compositeHeapBuffer(maxNumComponents));
+    }
+
+    @Override
+    public CompositeByteBuf compositeDirectBuffer(int maxNumComponents)
+    {
+        return track(super.compositeDirectBuffer(maxNumComponents));
+    }
+
+    private synchronized CompositeByteBuf track(CompositeByteBuf byteBuf)
+    {
+        trackedBuffers.add(byteBuf);
+        return byteBuf;
+    }
+
+    private synchronized ByteBuf track(ByteBuf byteBuf)
+    {
+        trackedBuffers.add(byteBuf);
+        return byteBuf;
+    }
+
+    private void checkInstanceForLeaks()
+    {
+        long referencedBuffersCount = 0;
+        synchronized (this) {
+            referencedBuffersCount = trackedBuffers.stream()
+                    .filter(byteBuf -> byteBuf.refCnt() > 0)
+                    .count();
+            // Make tracked buffers eligible for GC
+            trackedBuffers.clear();
+        }
+        // Trigger a GC. This will hopefully (but not necessarily) print details
+        // about detected leaks to standard error before the error is thrown.
+        System.gc();
```

Comment:
I think you want this inside of the if statement below, since it is only needed if there is a leak.